src/util/flatRender.js

```diff
@@ -0,0 +1,193 @@
+
+  // iterate subthoughts
+  return subThoughts.reduce((acc, child) => {
+    const childPath = unroot(startingPath.concat(child))
+    const value = child.value
+
+    // isParentCursorAncestor is used to prevent calling isDescendant everytime
+    // if the parent thought is already not an ancestor of the cursor then we don't need to call it everytime for its descendants
+    const isCursorAncestor =
+      isParentCursorAncestor &&
+      isDescendant(pathToContext(childPath), pathToContext(cursor))
+    const isCursor = equalPath(cursor, childPath)
+    const childPathLength = childPath.length
+
+    const children = getThoughtsRanked(childPath)
+
+    // decide if it is a distant ancestor that needs to be visible but needs to stop deeper recursion
+    const addDistantAncestorAndStop =
+      cursor.length - childPathLength <= (isLeaf ? 1 : 0) &&
+      !isCursor &&
+      !isCursorAncestor &&
+      !isCursorChildren
+    // stop recursion if distant ancestor doesn't need to be added to the array
+    const showDistantAncestor = !(
+      childPathLength < cursor.length &&
+      !isCursorAncestor &&
+      !addDistantAncestorAndStop
+    )
+
+    const isHidden = children.some(child => child.value === '=hidden')
+    const isMeta = isFunction(value)
+
+    // hide if this node is itself a meta function or has children meta =hidden
+    // if showHiddenThoughts is true then don't hide at all
+    const shouldHide =
+      !showHiddenThoughts &&
+      (isMeta || isHidden)
+
+    const depthInfo = {
+      hiddenNodes: acc.depthInfo.hiddenNodes + (isHidden ? 1 : 0),
+      metaNodes: acc.depthInfo.metaNodes + (isMeta ? 1 : 0),
+    }
+
+    if (!showDistantAncestor || shouldHide) return { ...acc, depthInfo }
+    else {
+      // siblings count which are not meta functions
+      const siblingsCountWithoutMeta = subThoughts.reduce(
+        (acc, child) => acc + (!isFunction(child.value) ? 1 : 0),
+        0
+      )
+
+      // stop deeper recursion at certain depth where any descendant of cursor has more than one subthought
+      // stop further deeper recursion if max depth is reached
+      const stop =
+        addDistantAncestorAndStop ||
+        (isCursorChildren && siblingsCountWithoutMeta > 1) ||
+        childPath.length - cursor.length === MAX_DEPTH_FROM_CURSOR
+
+      const distanceFromCursor = cursor.length - childPath.length
+
+      // if true the node will have reduced opacity on render
+      const isDistantThought =
+        (!isLeaf
+          ? distanceFromCursor >= 0
+          : distanceFromCursor >= (isCursorAncestor ? 2 : 1)) && !isCursor
+
+      const recursiveUpdate = getFlatArray({
+        startingPath: childPath,
+        children,
+        cursor,
+        isLeaf,
+        showHiddenThoughts,
+        isParentCursorAncestor: isCursorAncestor,
+        isCursorChildren: isCursorChildren || isCursor,
+      })
+
+      const deeperFlatArray = stop
+        ? []
+        : recursiveUpdate.flatArray
```

Comment:
Yes. It doesn't short circuit the logic right away. It gets to next recursion and stops. It is because `depthInfo` logic needs to calculate hidden nodes.

Reply:
I think we need to find a better way to do this. If `stop` is true, we shouldn't call `getFlatArray` again. We should just get the `depthInfo`. Note that if we were not passing along `depthInfo` for performance, there would be no reason to call `getFlatArray`. A performance optimization should not add unnecessary iterations due to new implementation details. 

It might seem messy to have to calculate `depthInfo` separately from `getFlatArray`, but that is the cost of introducing the additional complexity of this optimization. You are piggybacking on `getFlatArray` to get the `depthInfo`, but when `getFlatArray` isn't called, you will need to get `depthInfo` another way.

Reply:
I totally agree with you! It seems messy. We can either use nested iteration or we can use `depthInfo` logic to find if any children nodes are visible. To prevent extra level of recursion call with `depthInfo` logic, we calculate visible nodes by nested iteration when a node comes to stop instead of calling  `getFlatArray` again . Basically we don't go into extra recursion and do nested iterations only at leaf nodes. What do you think we should do ? Please let me know

Reply:
That sounds right.